### PR TITLE
Enable get instance by a String key

### PR DESCRIPTION
**Pull request by davidsdearaujo**
## What does this PR do
It creates a way to add and get binds by a **String** key.

Now **all** the **add** and **get** methods have a `String? key` parameter.

> **Search functions:** 
> When [key] is provided it will search the instance that have the same key

> **Add functions:**
> When [key] is provided this instance only can be found by key

## Changed Functions:
```dart
  T get<T>({ParamTransform? transform, String? key});

  T call<T>({ParamTransform? transform, String? key});

  void add<T>(Function constructor, {BindConfig<T>? config, String? key});

  void addInstance<T>(T instance, {BindConfig<T>? config, String? key});

  void addSingleton<T>(Function constructor, { BindConfig<T>? config, String? key});

  void addLazySingleton<T>(Function constructor, {BindConfig<T>? config, String? key});

  dynamic getNotifier<T>({String? key});

  T? tryGet<T>({ParamTransform? transform, String? key});

  bool isAdded<T>({String? key});

  bool isInstantiateSingleton<T>({String? key});

  T? disposeSingleton<T>({String? key});

  void replaceInstance<T>(T instance, {String? key});
```

### CODE REVIEW

File: lib/src/bind.dart [hunk 1]
```diff
@@ -42,24 +43,27 @@ class Bind<T> {
...
     final constructorListString = constructorString.split(' ');
     final paramsString = constructorListString[0];
-    final classNameString = constructorListString.last;
+    final classNameString = _resolveClassName<T>(constructorString);
+
     if (paramsString == '()') {
       return Bind<T>._(
         constructor: constructor,
-        className: classNameString,
+        className: (key != null) ? null : classNameString,
```
Comment (jacobaraujo7) [inline lib/src/bind.dart]:
```suggestion
        className: key ?? classNameString,
```

**Review by jacobaraujo7 — CHANGES_REQUESTED** [review]
Can you add some tests about this case?

**Comment by toshiossada** [comment]
LGTM

**Comment by davidsdearaujo** [comment]
> @jacobaraujo7
> Can you add some tests about this case?

The entire [auto_injector_tag_test.dart](https://github.com/Flutterando/auto_injector/pull/13/files#diff-a571f67bacf291119af78defb8dedc61efa78355ad32c1ad7bf2de4ced3939ac) is testing it!

**Comment by davidsdearaujo** [comment]
> @jacobaraujo7
> Can you add some tests about this case?

The entire [test/src/auto_injector_with_key_test.dart](https://github.com/Flutterando/auto_injector/pull/13/files#diff-61add6ae205229a44b6a4e064ecb7a8f3101c3a17570422f75786721c6991fc2) is testing it!

**Review by toshiossada — COMMENTED** [review]
LGFM

**Review by jacobaraujo7 — APPROVED** [review]

**Pull request merged**